### PR TITLE
Add unit tests for SA1012

### LINKED ISSUE #198: Diagnostic unit tests: SA1012 (OpeningCurlyBracketsMustBeSpacedCorrectly)
Implement unit tests for #18.

### PR DESCRIPTION
**Pull request by Noryoko**
If there is a better way to test the `({` scenario in the last test I'm open to suggestions.

fixes #198

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        [Fact]
+        public async Task TestParenthesisAsync()
+        {
+            var testCode = @"namespace TestNamespace
+{
+    public class TestClass
+    {
+        public void TestMethod()
+        {
+            TestMethod({ });
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs]: :question: Is it even possible to have `({` in a valid source file?

:bulb: If you only need to test `})`, you can use `(new object[] { })`.

**Comment by Noryoko** [comment]
I was coming to the same conclusion about `({`.

`})` Is covered by SA1013. I can remove that final test and refactor the analyzer to not check for `({`.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            var testCode = @"namespace TestNamespace
+{
+    public class TestClass
+    {
+        public void TestMethod()
+        {
+            var test = 2;
+            var x = $""{test}"";
+            x = $"" {test}"";
+            x = $"" { test}"";
```
Comment (Noryoko) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs]: Currently no diagnostics are returned inside of string interpolation. Should we report one if a space exists after the opening?

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs]: If you could do that, that would be *awesome*.

Comment (Noryoko) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1012UnitTests.cs]: Sure thing :)

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1012OpeningCurlyBracketsMustBeSpacedCorrectly.cs [hunk 3]
```diff
@@ -78,56 +78,32 @@ private void HandleOpenBraceToken(SyntaxTreeAnalysisContext context, SyntaxToken
...
-            {
-                precededBySpace = true;
-                allowLeadingSpace = true;
-                allowLeadingNoSpace = true;
-            }
-            else
+            if (!firstInLine)
             {
                 SyntaxToken precedingToken = token.GetPreviousToken();
                 precededBySpace = precedingToken.HasTrailingTrivia;
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1012OpeningCurlyBracketsMustBeSpacedCorrectly.cs]: :exclamation: `token.IsPrecededByWhitespace()` should be used here

**Comment by Noryoko** [comment]
The async usage analyzer alpha 2 was causing issues, merged master back in to get the updated analyzer

**Comment by Noryoko** [comment]
Looks like it can be simplified to `bool precededBySpace = token.IsFirstInLine() || token.IsPrecededByWhitespace();` and change the if statement to `if (!precededBySpace)`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1012OpeningCurlyBracketsMustBeSpacedCorrectly.cs [hunk 4]
```diff
@@ -78,56 +78,26 @@ private void HandleOpenBraceToken(SyntaxTreeAnalysisContext context, SyntaxToken
...
-                else if (!allowLeadingNoSpace && !precededBySpace)
-                {
-                    // Opening curly bracket must{} be {preceded} by a space.
-                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, token.GetLocation(), string.Empty, "preceded"));
-                }
+                // Opening curly bracket must{} be {preceded} by a space.
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, token.GetLocation(), string.Empty, "preceded"));
             }
 
             if (!lastInLine && !followedBySpace)
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1012OpeningCurlyBracketsMustBeSpacedCorrectly.cs]: `if (!token.IsLastInLine() && !followedBySpace)`
and remove it from line # 95

**Comment by Maxwe11** [comment]
>The async usage analyzer alpha 2 was causing issues, merged master back in to get the updated analyzer

For avoiding such situations you should keep your fork up to date.
Just do

    git remote add upstream https://github.com/DotNetAnalyzers/StyleCopAnalyzers.git

and do pull / fetch from upstream before start working.

**Comment by sharwell** [comment]
> For avoiding such situations you should keep your fork up to date

@Maxwe11 #1016 (which fixed this issue) was created *after* this pull request.

**Comment by Noryoko** [comment]
The updates happened today after I was already working on this branch.

**Comment by Maxwe11** [comment]
:+1:
Could you squash into two commits? One for test and one for diagnostic.

**Comment by sharwell** [comment]
@Maxwe11 I didn't see a pressing need to squash. I'd rather have it merged and let @Noryoko work on the next diagnostic. :smile:

**Pull request merged**